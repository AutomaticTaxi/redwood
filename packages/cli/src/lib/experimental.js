import path from 'path'

import { getConfig, getPaths, resolveFile } from '@redwoodjs/project-config'

/**
 * Checks to see if the user has the experimental server file present at api/src/server.ts|js and enabled in redwood.toml
 *
 * @returns {boolean} true if the experimental server file is enabled and present
 */
export function isUsingExperimentalServerFile() {
  return (
    getConfig().experimental?.serverFile?.enabled &&
    resolveFile(path.join(getPaths().api.src, 'server'))
  )
}
