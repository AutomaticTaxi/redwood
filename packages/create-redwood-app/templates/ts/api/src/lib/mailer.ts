import { Mailer } from '@redwoodjs/mailer-core'
import { InMemoryMailHandler } from '@redwoodjs/mailer-handler-in-memory'
import { NodemailerMailHandler } from '@redwoodjs/mailer-handler-nodemailer'
import { StudioMailHandler } from '@redwoodjs/mailer-handler-studio'
import { ReactEmailRenderer } from '@redwoodjs/mailer-renderer-react-email'

import { logger } from './logger'

export const mailer = new Mailer({
  // Handlers to send our emails out to the world
  handling: {
    handlers: {
      inMemory: new InMemoryMailHandler(),
      studio: new StudioMailHandler(),
      // TODO: Update this handler config or switch it out for a different handler completely
      nodemailer: new NodemailerMailHandler({
        transport: {
          host: 'localhost',
          port: 4319,
          secure: false,
        },
      }),
    },
    // This is our default handler when in production only
    default: 'nodemailer',
  },

  // Renderers to convert our templates into HTML/text
  rendering: {
    renderers: {
      reactEmail: new ReactEmailRenderer(),
    },
    default: 'reactEmail',
  },

  // Specific handlers for testing and development time
  test: {
    handler: 'inMemory',
    // when: process.env.NODE_ENV === 'test',
  },
  development: {
    handler: 'studio',
    // when: process.env.NODE_ENV !== 'production',
  },

  // Default send options
  defaults: {
    // replyTo: 'noreply@example.com',
  },

  logger,
})
