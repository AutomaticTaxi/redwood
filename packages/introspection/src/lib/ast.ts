import fs from 'node:fs'
import path from 'node:path'

import { parse as babelParse } from '@babel/parser'
import type { ParserPlugin } from '@babel/parser'
import traverse from '@babel/traverse'
import type {
  JSXElement,
  ExportNamedDeclaration,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/project-config'

export function getASTFromFile(filepath: string) {
  const code = fs.readFileSync(filepath, { encoding: 'utf8', flag: 'r' })

  // use jsx plugin for web files, because in JS, the .jsx extension is not used
  const isJsxFile =
    path.extname(filepath).match(/[jt]sx$/) ||
    path.parse(filepath).dir.startsWith(getPaths().web.base)

  const plugins = [
    'typescript',
    'nullishCoalescingOperator',
    'objectRestSpread',
    isJsxFile && 'jsx',
  ].filter(Boolean) as ParserPlugin[]

  return babelParse(code, {
    sourceType: 'module',
    plugins,
  })
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(element: JSXElement) {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
