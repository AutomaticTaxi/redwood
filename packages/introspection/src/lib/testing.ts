import { ensurePosixPath } from '@redwoodjs/project-config'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
