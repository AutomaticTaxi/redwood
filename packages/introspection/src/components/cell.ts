import fs from 'node:fs'
import path from 'node:path'

import traverse from '@babel/traverse'
import type { ExportNamedDeclaration } from '@babel/types'
import {
  isVariableDeclaration,
  isVariableDeclarator,
  isTaggedTemplateExpression,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/project-config'

import { getASTFromFile } from '../lib/ast'
import { getGraphQLQueryName } from '../lib/gql'

import type {
  RedwoodIntrospectionError,
  RedwoodIntrospectionWarning,
} from './introspection'
import { RedwoodIntrospectionComponent } from './introspection'

export class RedwoodCell extends RedwoodIntrospectionComponent {
  readonly type = 'cell'

  readonly gqlQuery: string | undefined
  readonly gqlQueryName: string | undefined

  readonly hasQueryExport: boolean
  readonly hasBeforeQueryExport: boolean
  readonly hasIsEmptyExport: boolean
  readonly hasAfterQueryExport: boolean
  readonly hasLoadingExport: boolean
  readonly hasEmptyExport: boolean
  readonly hasFailureExport: boolean
  readonly hasSuccessExport: boolean

  readonly isValid: boolean

  private constructor(filepath: string) {
    super(filepath)

    const ast = getASTFromFile(this.filepath)

    const namedExports: ExportNamedDeclaration[] = []
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        namedExports.push(path.node)
      },
    })

    const namedExportsNames = namedExports
      .map((node) => {
        if (
          isVariableDeclaration(node.declaration) &&
          isVariableDeclarator(node.declaration.declarations[0]) &&
          isIdentifier(node.declaration.declarations[0].id)
        ) {
          return node.declaration.declarations[0].id.name
        }
        return false
      })
      .filter(Boolean)

    // Check for a cells expected export fields
    this.hasQueryExport = namedExportsNames.includes('QUERY')
    this.hasBeforeQueryExport = namedExportsNames.includes('beforeQuery')
    this.hasIsEmptyExport = namedExportsNames.includes('isEmpty')
    this.hasAfterQueryExport = namedExportsNames.includes('afterQuery')
    this.hasLoadingExport = namedExportsNames.includes('Loading')
    this.hasEmptyExport = namedExportsNames.includes('Empty')
    this.hasFailureExport = namedExportsNames.includes('Failure')
    this.hasSuccessExport = namedExportsNames.includes('Success')

    const queryExport = namedExports.find((node) => {
      if (
        node.declaration != null &&
        isVariableDeclaration(node.declaration) &&
        isVariableDeclarator(node.declaration.declarations[0]) &&
        isIdentifier(node.declaration.declarations[0].id)
      ) {
        return node.declaration.declarations[0].id.name === 'QUERY'
      }
      return false
    })

    if (
      queryExport !== undefined &&
      queryExport.declaration != null &&
      isVariableDeclaration(queryExport.declaration) &&
      isVariableDeclarator(queryExport.declaration.declarations[0]) &&
      queryExport.declaration.declarations[0].init != null &&
      isTaggedTemplateExpression(
        queryExport.declaration.declarations[0].init
      ) &&
      isIdentifier(queryExport.declaration.declarations[0].init.tag) &&
      queryExport.declaration.declarations[0].init.tag.name === 'gql'
    ) {
      this.gqlQuery =
        queryExport.declaration.declarations[0].init.quasi.quasis[0].value.raw
      this.gqlQueryName = getGraphQLQueryName(this.gqlQuery)
    }

    // Determine if the cell is valid
    this.isValid = this.getErrors().length === 0
  }

  getErrors(): RedwoodIntrospectionError[] {
    const errors: RedwoodIntrospectionError[] = []

    // Check if the mandatory success export is present
    if (!this.hasSuccessExport) {
      errors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'No "Success" export found but one is required',
      })
    }

    // Check if the mandatory query export is present
    if (!this.hasQueryExport) {
      errors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'No "QUERY" export found but one is required',
      })
    }

    return errors
  }

  getWarnings(): RedwoodIntrospectionWarning[] {
    const warnings: RedwoodIntrospectionWarning[] = []

    // Check if the cell query could be extracted
    if (this.gqlQuery === undefined) {
      warnings.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'Unable to extract your cell query',
      })
    }

    // Check if the cell query is named
    if (this.gqlQueryName === undefined) {
      warnings.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'We recommend that you name your query operation',
      })
    }

    return warnings
  }

  static parseCell(filepath: string) {
    return new RedwoodCell(filepath)
  }

  static parseCells(directory: string = getPaths().web.components) {
    const cells: RedwoodCell[] = []

    // TODO: Confirm this the condition to detect a cell
    // Cells must be defined within a file which ends with `Cell.{js, jsx, tsx}`
    const getCellFiles = (directory: string) => {
      const cellFiles: string[] = []

      const directoryContents = fs.readdirSync(directory)
      directoryContents.forEach((content) => {
        const stat = fs.lstatSync(path.join(directory, content))
        if (stat.isDirectory()) {
          cellFiles.push(...getCellFiles(path.join(directory, content)))
        } else if (stat.isFile()) {
          if (content.match(/.+Cell\.(js|jsx|tsx)$/)) {
            cellFiles.push(path.join(directory, content))
          }
        }
      })

      return cellFiles
    }

    const cellFiles = getCellFiles(directory)
    cellFiles.forEach((cellPath) => {
      cells.push(this.parseCell(cellPath))
    })

    return cells
  }
}
