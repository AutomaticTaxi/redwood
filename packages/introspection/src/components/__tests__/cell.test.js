import path from 'node:path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodCell } from '../cell'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)
const FIXTURE_PROJECTS = [
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
]

describe.each(FIXTURE_PROJECTS)(
  "From within the '%s' fixture",
  (PROJECT_NAME) => {
    // Setup RWJS_CWD to point to the fixture project
    const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)
    const RWJS_CWD = process.env.RWJS_CWD
    beforeAll(() => {
      process.env.RWJS_CWD = PROJECT_PATH
    })
    afterAll(() => {
      process.env.RWJS_CWD = RWJS_CWD
    })

    it('returns the correct cells', () => {
      const cells = RedwoodCell.parseCells()
      expect(cells.length).toMatchSnapshot('cells count')
      for (let i = 0; i < cells.length; i++) {
        const cell = cells[i]
        cell.filepath = stripAndFormatPathForTesting(
          cell.filepath,
          PROJECT_PATH
        )
        expect(cell).toMatchSnapshot(`${cell.filepath} content`)
        expect(cell.getErrors()).toMatchSnapshot(`${cell.filepath} errors`)
        expect(cell.getWarnings()).toMatchSnapshot(`${cell.filepath} warnings`)
      }
    })
  }
)
