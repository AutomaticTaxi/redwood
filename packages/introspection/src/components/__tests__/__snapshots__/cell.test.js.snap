// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the 'empty-project' fixture returns the correct cells: cells count 1`] = `0`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosCell/NumTodosCell.js content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/NumTodosCell/NumTodosCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosCell/NumTodosCell.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosCell/NumTodosCell.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosTwoCell/NumTodosTwoCell.js content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/NumTodosTwoCell/NumTodosTwoCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosTwoCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosTwoCell/NumTodosTwoCell.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/NumTodosTwoCell/NumTodosTwoCell.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js errors 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "No "Success" export found but one is required",
  },
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "No "QUERY" export found but one is required",
  },
]
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js warnings 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "Unable to extract your cell query",
  },
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "We recommend that you name your query operation",
  },
]
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/TodoListCell/TodoListCell.tsx",
  "gqlQuery": "
  query TodoListCell_GetTodos {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": "TodoListCell_GetTodos",
  "hasAfterQueryExport": true,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.tsx errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.tsx warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct cells: cells count 1`] = `4`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js errors 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "No "Success" export found but one is required",
  },
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "No "QUERY" export found but one is required",
  },
]
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TableCell/TableCell.js warnings 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "Unable to extract your cell query",
  },
  {
    "component": {
      "filepath": "/web/src/components/TableCell/TableCell.js",
      "name": "TableCell",
      "type": "cell",
    },
    "message": "We recommend that you name your query operation",
  },
]
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.js content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/TodoListCell/TodoListCell.js",
  "gqlQuery": "
  {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "type": "cell",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: /web/src/components/TodoListCell/TodoListCell.js warnings 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/components/TodoListCell/TodoListCell.js",
      "name": "TodoListCell",
      "type": "cell",
    },
    "message": "We recommend that you name your query operation",
  },
]
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct cells: cells count 1`] = `2`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/AuthorCell/AuthorCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/AuthorCell/AuthorCell.tsx",
  "gqlQuery": "
  query FindAuthorQuery($id: Int!) {
    author: user(id: $id) {
      email
      fullName
    }
  }
",
  "gqlQueryName": "FindAuthorQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "AuthorCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/AuthorCell/AuthorCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/AuthorCell/AuthorCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostCell/BlogPostCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/BlogPostCell/BlogPostCell.tsx",
  "gqlQuery": "
  query FindBlogPostQuery($id: Int!) {
    blogPost: post(id: $id) {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "FindBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostCell/BlogPostCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostCell/BlogPostCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostsCell/BlogPostsCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/BlogPostsCell/BlogPostsCell.tsx",
  "gqlQuery": "
  query BlogPostsQuery {
    blogPosts: posts {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "BlogPostsQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostsCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostsCell/BlogPostsCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/BlogPostsCell/BlogPostsCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactCell/ContactCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Contact/ContactCell/ContactCell.tsx",
  "gqlQuery": "
  query FindContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactCell/ContactCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactCell/ContactCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactsCell/ContactsCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Contact/ContactsCell/ContactsCell.tsx",
  "gqlQuery": "
  query FindContacts {
    contacts {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContacts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactsCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactsCell/ContactsCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/ContactsCell/ContactsCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/EditContactCell/EditContactCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Contact/EditContactCell/EditContactCell.tsx",
  "gqlQuery": "
  query EditContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "EditContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditContactCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/EditContactCell/EditContactCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Contact/EditContactCell/EditContactCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/EditPostCell/EditPostCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Post/EditPostCell/EditPostCell.tsx",
  "gqlQuery": "
  query EditPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "EditPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditPostCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/EditPostCell/EditPostCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/EditPostCell/EditPostCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostCell/PostCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Post/PostCell/PostCell.tsx",
  "gqlQuery": "
  query FindPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostCell/PostCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostCell/PostCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostsCell/PostsCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/Post/PostsCell/PostsCell.tsx",
  "gqlQuery": "
  query FindPosts {
    posts {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPosts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostsCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostsCell/PostsCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/Post/PostsCell/PostsCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx content 1`] = `
RedwoodCell {
  "filepath": "/web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx",
  "gqlQuery": "
  query FindWaterfallBlogPostQuery($id: Int!) {
    waterfallBlogPost: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindWaterfallBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "WaterfallBlogPostCell",
  "type": "cell",
}
`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: /web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct cells: cells count 1`] = `10`;
